### PR TITLE
Another implementation of CRF

### PR DESCRIPTION
**Pull request by linxihui**
This implementation has two modes for optimization:
    1. (`join mode`) optimized by maximizing join likelihood, which is optimal in theory of statistics.
       Note that in this case, CRF must be the output/last layer.
    2. (`marginal mode`) return marginal probabilities on each time step and optimized via composite
       likelihood (product of marginal likelihood), i.e., using `categorical_crossentropy` loss.
       Note that in this case, CRF can be either the last/output layer or an intermediate layer (not clear now whether there is a valuable application).

  For prediction (test phrase), one can choose either Viterbi best path (class indices) or marginal
probabilities if probabilities are needed. 

   This implementation also supports masking, specifying activation functions (or more accurate, restrict range of chain energy) and accepts either onehot or sparse target.

NOTE: this PR requires PR https://github.com/fchollet/keras/pull/4899

### CODE REVIEW

**Comment by EderSantana** [comment]
would you please add unit test and an example? New users would benefit with a good use case to know this layer exists. It would also help to know how to properly apply it.

**Comment by linxihui** [comment]
@EderSantana yup, will do it soon. I made this just because there is another PR (https://github.com/fchollet/keras/pull/4621) for CRF, and I think it may be valuable to compare.

**Comment by linxihui** [comment]
@EderSantana @fchollet  Sorry for the delay. Unit tests and an example is added.  The check fail because of some other tests [`test_optimizers.py`](https://travis-ci.org/fchollet/keras/jobs/188614457#L1003). Can you take a look? Also, I find that quite a few other tests in the master branch fail, mostly `convolution` layers.

**Comment by EderSantana** [comment]
the failure is only in theano under python 2.7 and unrelated to this current PR
@fchollet did anything change there recently?

**Comment by linxihui** [comment]
@EderSantana it's weird. Now it passes tests.  Though there is a seed setting in `test_optimizers.py`, some other unexpected randomness might exist.

File: keras/layers/crf.py [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+
+    # Examples
+
+    ```python
+        model = Sequential()
+        model.add(Embedding(3001, 300, mask_zero=True)(X)
+
+        # use learn_mode = 'join', test_mode = 'viterbi'
+        crf = CRF(10)
+        model.add(crf(Embed))
```
Comment (himkt) [inline keras/layers/crf.py]: hi @linxihui, 'Embed' seems to be undefined.
I think this line should be ```model.add(crf)``` (followed by [test_crf.py#L34](https://github.com/linxihui/keras/blob/7e8495c5fe4597374ff7b84afe5bbffc9692de75/tests/keras/layers/test_crf.py#L34))

Comment (linxihui) [inline keras/layers/crf.py]: @himkt You're right. Thanks for pointing out. Now fixed.

**Comment by fchollet** [comment]
Closing outdated PR. If you still care about the content of the PR, please submit a new PR to `master`, updated for the Keras 2.0 API.

**Comment by himkt** [comment]
@linxihui Is there any plan about this PR in future ? I think this PR is great work (I wanna use CRF!). I want to help if there are something I can support.

**Comment by linxihui** [comment]
@himkt  See https://github.com/fchollet/keras/pull/6226

**Pull request closed without merging**